Vault env variables

notlee opened the pull request:
Adds a `make .env` command to update a local `.env` file with environment variables from Vault. 

These can be accessed through the [web ui](https://vault-ui.in.ft.com/secrets/generic/secret/teams/origami/origami-registry-ui/local) or via the vault cli. See the [getting started guide](https://github.com/Financial-Times/vault/wiki/Getting-Started).

Vault secrets are stored in a directory structure. The structure I've gone for is `origami/${systemCode}/${region}/`. E.g. `origami/origami-registry-ui/local`.

**Suggested Next steps:**

Use Vault for all env storage/retrieval.
- Push all secrets to Vault (local and otherwise).
- Update services docs to call `make .env` after `make install` for local development.
- Update documentation to point to Vault rather than LastPass to retrieve production environment variables.
- Remove secrets from LastPass..?

That's a safe milestone with secrets manageable outside of LastPass and auto .env population to get up and running locally. We may then follow up by enabling qa/production environment variables to update automatically, perhaps using a [vault webhook](https://github.com/Financial-Times/vault/wiki/Webhooks) along the lines of [next-vault-sync](https://github.com/Financial-Times/next-vault-sync)

Commit:
Add `.env` to update local environment variables from Vault.

Commit:
Pass region env through to the `.env` target.

JakeChampion commented on index.mk:
```diff
@@ -36,6 +36,15 @@ ci: verify test whitesource
+.env:
+	@if [[ -z "$(shell command -v vault)" ]]; then echo "Error: You don't have Vault installed. Follow the guide at https://github.com/Financial-Times/vault/wiki/Getting-Started"; exit 1; fi
+	@if [[ -z "$(shell find ~/.vault-token -mmin -480)" ]]; then echo "Error: You are not logged into Vault. Try vault auth --method github."; exit 1; fi
+	@if [[ -z "$(shell grep .env .gitignore)" ]]; then echo "Error: .gitignore must include .env"; exit 1; fi
```
We should also check if `.env` exists in `.npmignore` because `npm publish` will not use `.gitignore` if a `.npmignore` exists.

JakeChampion commented on index.mk:
```diff
@@ -36,6 +36,15 @@ ci: verify test whitesource
+	@if [[ -z "$(shell command -v vault)" ]]; then echo "Error: You don't have Vault installed. Follow the guide at https://github.com/Financial-Times/vault/wiki/Getting-Started"; exit 1; fi
+	@if [[ -z "$(shell find ~/.vault-token -mmin -480)" ]]; then echo "Error: You are not logged into Vault. Try vault auth --method github."; exit 1; fi
+	@if [[ -z "$(shell grep .env .gitignore)" ]]; then echo "Error: .gitignore must include .env"; exit 1; fi
+	@SERVICE_SYSTEM_CODE=$(SERVICE_SYSTEM_CODE) REGION=$(REGION) node vault.js
```
Is this the first make task that requires NodeJS? If so, we should document which minimum version of NodeJS is required.

notlee replied:
Yeah it's the first. I avoided `Object.entries` thinking I needed to support Node.js 6.x, but actually [Node.js 8.x is required for origami-service](https://github.com/Financial-Times/origami-service#requirements).

I'm assuming this isn't a breaking change as we already assume the existence of npm.

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    // Get secrets object.
+    var existingSecrets = fs.readFileSync(envFile).toString().trim().split('\n');
```
Is there a reason this is a `var` and not a `let`?

JakeChampion replied:
Could turn into a `const`:
```
const existingSecrets = fs.readFileSync(envFile).toString().trim().split('\n').reduce((secretsObject, secret) => {
    const secretArr = (secret.includes('=') ? secret.split('=') : null);
    if (secretArr) {
        secretsObject[secretArr[0]] = secretArr[1];
    }
    return secretsObject;
}, {});
```

notlee replied:
copy 🍝

notlee commented on index.mk:
```diff
@@ -36,6 +36,15 @@ ci: verify test whitesource
+	@if [[ -z "$(shell command -v vault)" ]]; then echo "Error: You don't have Vault installed. Follow the guide at https://github.com/Financial-Times/vault/wiki/Getting-Started"; exit 1; fi
+	@if [[ -z "$(shell find ~/.vault-token -mmin -480)" ]]; then echo "Error: You are not logged into Vault. Try vault auth --method github."; exit 1; fi
+	@if [[ -z "$(shell grep .env .gitignore)" ]]; then echo "Error: .gitignore must include .env"; exit 1; fi
+	@SERVICE_SYSTEM_CODE=$(SERVICE_SYSTEM_CODE) REGION=$(REGION) node vault.js
```
whoops

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,56 @@
+        return secretsObject;
+    }, {});
+    const vaultSecrets = result.data;
+    const secrets = Object.assign(existingSecrets, vaultSecrets)
```
This would overwrite an existing secret with a secret of the same name from vault, is this the expected behaviour?

notlee replied:
That's what I intended, expecting Vault to be the source of truth. I did not want to remove an environment variable not configured in vault though.

rowanmanning replied:
I think this is what we want, yeah. One addition that might be useful though is if the script warns when a secret is found locally but not in Vault – there's a chance it _should_ be in Vault.

notlee replied:
Good suggestion, I've added that.

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,56 @@
+    }
+
+    // Write .env secrets.
+    fs.writeFile(envFile, envContent, () => {
```
I think we are swallowing errors here due to the lack of taking in an error parameter in the callback. Considering other parts of this scripts are using synchronous file system calls, we should make this one also synchronous, that will also stop the error being swallowed.

Commit:
Call node vault script with the correct path. :mask:

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,56 @@
+    // Get secrets object.
+    var existingSecrets = fs.readFileSync(envFile).toString().trim().split('\n');
+    existingSecrets = existingSecrets.reduce((secretsObject, secret) => {
+        const secretArr = (secret.includes('=') ? secret.split('=') : null);
```
This looks to remove comments that may exist in `.env` files. Comments start with `#`.

notlee replied:
Good thinking. That is an inconvenient thought.

notlee replied:
I've made an updated to keep comments which are on their own line. However comments which are inline are still removed if their value is updated from Vault.

Commit:
Remove `var` use. Use fs.write and output errors.

Commit:
Keep .env comments which are on their own line.

Commit:
Use writeFileSync correctly

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,65 @@
+const fs = require('fs');
+
+const service = process.env.SERVICE_SYSTEM_CODE;
+let region = (process.env.REGION || 'LOCAL').toLowerCase();
```
This can also be `const`

Commit:
Warn of enviroment variables which are not stored in Vault.

Commit:
Correct `let` use to `const`.

Commit:
Add Node requirement to the README.

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,71 @@
+    // Get secrets object.
+    const existingComments = {};
+    const existingSecrets = fs.readFileSync(envFile).toString().trim().split('\n').reduce((secretsObject, line, index) => {
+        if (line.trim().indexOf('#') == 0) {
```
You can use `String.prototype.startsWith('#')` instead of `String.prototype.indexOf('#') == 0`

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,71 @@
+        if (line.trim().indexOf('#') == 0) {
+            // Line is a comment.
+            existingComments[index] = line;
+        } else {
```
Perhaps we should try and detect inline comments and throw an error if they are detected. That way we won't remove the inline comments silently, they would need to be manually moved to a separate line.

I think we can detect this scenario with a regular expression like this -- https://regex101.com/r/2zxgvU/2

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,71 @@
+const { exec } = require('child_process');
```
Add `"use strict";` to the top of the file to ensure we are in strict mode, it will help find some bugs.

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,71 @@
+    // Create string of secrets for .env.
+    const keys = Object.keys(secrets);
+    let envContent = '';
+    for (index = 0; index <= keys.length - 1; index++) {
```
`index` is currently a global variable, update this to `let index = 0` and it will no longer be global scoped but block scoped

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,71 @@
+    const nonVaultSecretKeys = Object.keys(existingSecrets).filter((existingKey) => {
+        return !vaultSecretKeys.includes(existingKey);
+    });
+    console.log(`The following enviroment variables are custom and not stored in Vault:\n${nonVaultSecretKeys}.`);
```
spelling mistake for word `enviroment`

Commit:
Remove redundant(?) `sinclude` to prevent `.env` running by default.

Commit:
Consolidate PHONY

Commit:
Error if .npmignore does not include .env

Commit:
Spelling of environment.

Commit:
Remove global variable :scream:

Commit:
Use startsWith

Commit:
Exit with success. ".npmignore" is not required,

Commit:
Check for env inline comments and exit if they will be overridden.

JakeChampion commented on index.mk:
```diff
@@ -19,6 +18,12 @@ endif
 export PATH := $(PATH):$(NPM_BIN)
 
+ifeq ("$(wildcard node_modules/@financial-times/origami-service-makefile/index.mk)","")
+PATH_TO_SERVICE_MAKEFILE :=
```
What would happen if this condition executed?

notlee replied:
The env is set but empty. For our purposes though I don't think this is necessary seen as we don't check if it is set.

notlee replied:
So I'll update this:
```
ifneq ("$(wildcard node_modules/@financial-times/origami-service-makefile/index.mk)","")
PATH_TO_SERVICE_MAKEFILE := node_modules/@financial-times/origami-service-makefile/
endif
```

JakeChampion commented on index.mk:
```diff
@@ -36,6 +41,15 @@ ci: verify test whitesource
+.env:
+	@if [[ -z "$(shell command -v vault)" ]]; then echo "Error: You don't have Vault installed. Follow the guide at https://github.com/Financial-Times/vault/wiki/Getting-Started"; exit 1; fi
+	@if [[ -z "$(shell find ~/.vault-token -mmin -480)" ]]; then echo "Error: You are not logged into Vault. Try vault auth --method github."; exit 1; fi
+	@if [[ -z "$(shell grep .env .gitignore)" ]]; then echo "Error: .gitignore must include .env"; exit 1; fi
```
Should `--silent --no-messages` be added to the `grep` here like it is on line 49 below?

notlee replied:
The changes below below allow `npmignore` to be omitted. I think I'm happy to leave the output if `.gitignore` is not present -- with the assumption git is used.

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,89 @@
+if (!service) {
+    console.warn(
+        'Could not load environment variables from Vault. Service name is undefined.',
+        'Set the environment variable "SERVICE_SYSTEM_CODE" in your Makefile.'
```
What is a service name? Is it the same as a system code?

notlee replied:
Good spot (should be "code").

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,89 @@
+            (stdout ? `\nstdout: ${stdout}` : ''),
+            (stderr ? `\nstderr: ${stderr}` : '')
+        );
+        process.exit(1);
```
We should avoid using `process.exit(1)` and use `process.exitCode = 1;` instead.

Calling `process.exit(1)` forces the process to exit as quickly as possible, even if there are still asynchronous operations pending that have not yet completed fully. E.G. The `console.warn` on the line above may not be written to `process.stdout` because `console.warn` may sometimes be asynchronous.

notlee replied:
TIL! Thanks. I've replaced the with throwing an error.

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,89 @@
+
+    // Get secrets object.
+    const existingComments = {};
+    const existingSecrets = fs.readFileSync(envFile).toString().trim().split('\n').reduce((secretsObject, line, index) => {
```
What would happen if the file `.env` does not exist?

notlee replied:
Bad things. Not anymore.

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,89 @@
+        const inlineComments = /^\s*(?:"(?:.*?)"|'(?:.*?)'|(?:[^\s#]*))?\s*(#.*)/mg;
+        if (inlineComments.exec(replacedSecret)) {
+            console.log(`The "${replacedSecretKey}" enviroment variable has an inline comment. Please move this comment to its own line.`);
+            process.exit(1);
```
Should change to `process.exitCode = 1`

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,89 @@
+        return !vaultSecretKeys.includes(existingKey);
+    });
+    console.log(`The following environment variables are custom and not stored in Vault:\n${nonVaultSecretKeys}.`);
+    process.exit(0);
```
`process.exitCode = 0;`

JakeChampion commented on lib/vault.js:
```diff
@@ -0,0 +1,89 @@
+    const secrets = Object.assign({}, existingSecrets, vaultSecrets)
+    const keys = Object.keys(secrets);
+    let envContent = '';
+    for (let index = 0; index <= keys.length - 1; index++) {
```
If we had more comments than keys, we would lose some of the comments

JakeChampion reviewed (changes requested):
I think it should be extracted into it's own module with a test suite

Commit:
Remove use of `process.exit`'.

As it exits as quickly as possible even if there are still asynchronous operations pending (e.g. console.log).

Commit:
Handle .env not existings.

Commit:
Account for more comments than secrets (eww, in need of refactor)

notlee commented:
@JakeChampion Agreed, as it's grown more than expected I'll refactor and extract into its own module with tests.
